# refseq_bipype.py

PATH_16S_DATABASE = '/home/pszczesny/soft/bipype/SSU_candidate_db.fasta'
PATH_ITS_DATABASE = '/home/pszczesny/soft/bipype/UNITE_public_from_27.01.13.fasta'
GI_TAX_PATH = '/home/pszczesny/workingdata/refseq/db/tax_id/gi_taxid_nucl.dmp'
TAX_NAME_PATH = '/home/pszczesny/workingdata/refseq/db/tax_id/names.dmp'
RAP_LOC = '/home/pszczesny/soft/RAPSearch2.12_64bits/bin/rapsearch'
REF_PROT_LOC_MASL28 = '/home/pszczesny/storage/workingdata/rapsearch/masl28'
REF_PROT_LOC_KO = '/home/pszczesny/soft/KEGG/ko.pep.rapsearch.db'
REF_PROT_LOC_ELSE = '/home/pszczesny/workingdata/refseq/protein/refseq_protein'
USEARCH_LOC = '/home/pszczesny/soft/usearch'
FQ2FA_PATH = '/usr/local/bioinformatics/fastx_toolkit/bin/fastq_to_fasta'
HUMANN_PATH = '/home/pszczesny/soft/humann-0.99'
HUMANN_DATA_PATH = '/home/pszczesny/soft/humann-0.99/input/hmp_metadata.dat'

# bipype

REF_FUNGI = '/home/pszczesny/workingdata/refseq/db/fungi/fungi_refseq'
REF_PLANT_1 = '/home/pszczesny/workingdata/refseq/db/plant/plant_refseq_1'
REF_PLANT_2 = '/home/pszczesny/workingdata/refseq/db/plant/plant_refseq_2'
X16S_DB = '/public/EMIRGE/SSU/SSU_candidate_db.udb'
ITS_DB = '/home/pszczesny/workingdata/ITS/unite.udb'
NCBI_TAXA_DB = '/home/pszczesny/soft/bipype/NCBI_tax_id_name.pickle'
SSUS_TAXA_DB = '/home/pszczesny/soft/bipype/SSUs_tax_id_name.pickle'
RECONSTRUCT_DB = '/home/pszczesny/storage/workingdata/mcra/mcra10k.fasta'
